File: src/update-status.js
const fs = require("fs");
const path = require("path");
const os = require("os");

const STATUS_DIR = path.join(os.homedir(), ".claude", "status");
const SESSIONS_DIR = path.join(STATUS_DIR, "sessions");
const CURRENT_FILE = path.join(STATUS_DIR, "current.json");
const CLAUDE_SESSIONS_DIR = path.join(os.homedir(), ".claude", "sessions");

// Maps hook event name to status
const EVENT_STATUS = {
  SessionStart: "starting",
  UserPromptSubmit: "running",
  PreToolUse: "running",
  PostToolUseFailure: "error",
  PermissionRequest: "waiting_approval",
  Stop: "idle",
  SessionEnd: "ended",
};

function readStdin() {
  return new Promise((resolve, reject) => {
    let data = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (chunk) => (data += chunk));
    process.stdin.on("end", () => resolve(data));
    process.stdin.on("error", reject);
    setTimeout(() => resolve(data), 500);
  });
}

/**
 * Look up the session name (set via /rename) from ~/.claude/sessions/.
 * Files are named {pid}.json and contain sessionId + name.
 */
function findSessionName(sessionId) {
  if (!sessionId) return "";
  try {
    const files = fs.readdirSync(CLAUDE_SESSIONS_DIR);
    for (const file of files) {
      if (!file.endsWith(".json")) continue;
      try {
        const data = JSON.parse(
          fs.readFileSync(path.join(CLAUDE_SESSIONS_DIR, file), "utf8")
        );
        if (data.sessionId === sessionId) {
          return data.name || "";
        }
      } catch {}
    }
  } catch {}
  return "";
}

async function main() {
  const raw = await readStdin();
  let input = {};
  try {
    input = raw ? JSON.parse(raw) : {};
  } catch {}

  const event = input.hook_event_name || "";
  const status = EVENT_STATUS[event] || "running";
  const toolName = input.tool_name || "";
  const cwd = input.cwd || "";
  const sessionId = input.session_id || "";

  // Skip events without a valid session_id
  if (!sessionId || sessionId === "unknown") {
    process.exit(0);
  }

  // Look up session display name
  const sessionName = findSessionName(sessionId);

  // Read previous state for this session to preserve fields
  const sessionFile = path.join(SESSIONS_DIR, `${sessionId}.json`);
  let prev = {};
  try {
    prev = JSON.parse(fs.readFileSync(sessionFile, "utf8"));
  } catch {}

  // Build display name: {project}-{name} if renamed, else just {project}
  const project = cwd ? path.basename(cwd) : prev.project || "";
  let displayName;
  if (sessionName) {
    displayName = `${project}-${sessionName}`;
  } else if (prev.display_name && status !== "starting") {
    displayName = prev.display_name;
  } else {
    // No rename name: use project name only.
    // Deduplicate: if another session already uses this name, append a counter.
    displayName = project;
    try {
      const existing = fs.readdirSync(SESSIONS_DIR)
        .filter(f => f.endsWith(".json") && f !== `${sessionId}.json`);
      const usedNames = [];
      for (const ef of existing) {
        try {
          const ed = JSON.parse(fs.readFileSync(path.join(SESSIONS_DIR, ef), "utf8"));
          usedNames.push(ed.display_name || "");
        } catch {}
      }
      let candidate = displayName;
      let counter = 2;
      while (usedNames.includes(candidate)) {
        candidate = `${displayName} (${counter})`;
        counter++;
      }
      displayName = candidate;
    } catch {}
  }

  const entry = {
    session_id: sessionId,
    display_name: displayName,
    project: project || prev.project || "",
    status: status,
    tool: status === "idle" || status === "starting" || status === "ended"
      ? ""
      : toolName || prev.tool || "",
    cwd: cwd || prev.cwd || "",
    last_update: new Date().toISOString(),
  };

  // Write per-session file
  fs.mkdirSync(SESSIONS_DIR, { recursive: true });
  fs.writeFileSync(sessionFile, JSON.stringify(entry, null, 2));

  // Update active session pointer
  const current = {
    active_session: sessionId,
    last_update: new Date().toISOString(),
  };
  fs.writeFileSync(CURRENT_FILE, JSON.stringify(current, null, 2));

  process.exit(0);
}

main().catch(() => process.exit(0));

File: src/update-status.py
"""Claude Code hook — reads event JSON from stdin, writes status file.
Run with pythonw.exe to avoid console-window flash on Windows."""
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

STATUS_DIR = os.path.join(os.path.expanduser("~"), ".claude", "status")
SESSIONS_DIR = os.path.join(STATUS_DIR, "sessions")
CURRENT_FILE = os.path.join(STATUS_DIR, "current.json")
CLAUDE_SESSIONS_DIR = os.path.join(os.path.expanduser("~"), ".claude", "sessions")

EVENT_STATUS = {
    "SessionStart":       "starting",
    "UserPromptSubmit":   "running",
    "PreToolUse":         "running",
    "PostToolUseFailure": "error",
    "PermissionRequest":  "waiting_approval",
    "Stop":               "idle",
    "SessionEnd":         "ended",
}


def find_session_name(session_id: str) -> str:
    """Look up the session name (set via /rename) from ~/.claude/sessions/."""
    if not session_id:
        return ""
    try:
        for fname in os.listdir(CLAUDE_SESSIONS_DIR):
            if not fname.endswith(".json"):
                continue
            try:
                with open(os.path.join(CLAUDE_SESSIONS_DIR, fname),
                          "r", encoding="utf-8") as fh:
                    data = json.load(fh)
                if data.get("sessionId") == session_id:
                    return data.get("name", "")
            except Exception:
                pass
    except Exception:
        pass
    return ""


def main():
    # Read JSON from stdin (Claude Code pipes hook event data)
    try:
        raw = sys.stdin.read()
    except Exception:
        raw = ""

    try:
        inp = json.loads(raw) if raw else {}
    except Exception:
        inp = {}

    event = inp.get("hook_event_name", "")
    status = EVENT_STATUS.get(event, "running")
    tool_name = inp.get("tool_name", "")
    cwd = inp.get("cwd", "")
    session_id = inp.get("session_id", "")

    # Skip events without a valid session_id
    if not session_id or session_id == "unknown":
        return

    session_name = find_session_name(session_id)

    # Read previous state to preserve fields
    session_file = os.path.join(SESSIONS_DIR, f"{session_id}.json")
    prev = {}
    try:
        with open(session_file, "r", encoding="utf-8") as fh:
            prev = json.load(fh)
    except Exception:
        pass

    project = Path(cwd).name if cwd else prev.get("project", "")

    # Build display name
    if session_name:
        display_name = f"{project}-{session_name}"
    elif prev.get("display_name") and status != "starting":
        display_name = prev["display_name"]
    else:
        display_name = project
        # Deduplicate
        try:
            existing = [f for f in os.listdir(SESSIONS_DIR)
                        if f.endswith(".json") and f != f"{session_id}.json"]
            used = []
            for ef in existing:
                try:
                    with open(os.path.join(SESSIONS_DIR, ef),
                              "r", encoding="utf-8") as fh:
                        used.append(json.load(fh).get("display_name", ""))
                except Exception:
                    pass
            candidate = display_name
            counter = 2
            while candidate in used:
                candidate = f"{display_name} ({counter})"
                counter += 1
            display_name = candidate
        except Exception:
            pass

    entry = {
        "session_id": session_id,
        "display_name": display_name,
        "project": project or prev.get("project", ""),
        "status": status,
        "tool": ("" if status in ("idle", "starting", "ended")
                 else tool_name or prev.get("tool", "")),
        "cwd": cwd or prev.get("cwd", ""),
        "last_update": datetime.now(timezone.utc).isoformat(),
    }

    os.makedirs(SESSIONS_DIR, exist_ok=True)
    with open(session_file, "w", encoding="utf-8") as fh:
        json.dump(entry, fh, ensure_ascii=False, indent=2)

    current = {
        "active_session": session_id,
        "last_update": datetime.now(timezone.utc).isoformat(),
    }
    with open(CURRENT_FILE, "w", encoding="utf-8") as fh:
        json.dump(current, fh, ensure_ascii=False, indent=2)


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass
